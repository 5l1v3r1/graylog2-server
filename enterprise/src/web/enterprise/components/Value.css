:local(.dropdowntoggle) {
    cursor: pointer;
}

:local(.dropdowntoggle):hover {
    color: #3967ff;
    text-decoration: underline;
}
