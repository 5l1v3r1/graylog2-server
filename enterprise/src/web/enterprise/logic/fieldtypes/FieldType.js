import Immutable from 'immutable';

const Properties = {
  Compound: 'compound',
  Enumerable: 'enumerable',
  FullTextSearch: 'full-text-search',
  Numeric: 'numeric',
};

class FieldType {
  constructor(type, properties, indexNames) {
    this.value = Immutable.Map({ type, properties: Immutable.Set(properties), indexNames: Immutable.Set(indexNames) });
  }

  static Unknown = new FieldType('unknown', [], []);

  get type() {
    return this.value.get('type');
  }

  get properties() {
    return this.value.get('properties');
  }

  get indexNames() {
    return this.value.get('indexNames');
  }

  isNumeric() {
    return this.properties.has(Properties.Numeric);
  }

  isCompound() {
    return this.properties.has(Properties.Compound);
  }

  static fromJSON(value) {
    // eslint-disable-next-line camelcase
    const { type, properties, index_names } = value;
    return new FieldType(type, properties, index_names);
  }
}

export default FieldType;
