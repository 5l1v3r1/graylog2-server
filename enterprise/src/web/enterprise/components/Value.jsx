import React from 'react';
import PropTypes from 'prop-types';
import { PluginStore } from 'graylog-web-plugin/plugin';
import { MenuItem, Well } from 'react-bootstrap';

import UserTimezoneTimestamp from 'enterprise/components/common/UserTimezoneTimestamp';
import FieldType from 'enterprise/logic/fieldtypes/FieldType';

import OverlayDropdown from './OverlayDropdown';
import style from './Value.css';
import CustomPropTypes from './CustomPropTypes';

export default class Value extends React.Component {
  static propTypes = {
    children: PropTypes.node,
    field: PropTypes.string.isRequired,
    menuContainer: PropTypes.object,
    queryId: PropTypes.string.isRequired,
    type: CustomPropTypes.FieldType,
    value: PropTypes.node.isRequired,
  };

  static defaultProps = {
    type: null,
    children: null,
    interactive: false,
    viewId: null,
    menuContainer: document.body,
    type: FieldType.Unknown,
  };

  constructor(props, context) {
    super(props, context);
    this.state = {
      open: false,
    };
  }

  _onMenuToggle = () => this.setState(state => ({ open: !state.open }));
  _renderTypeSpecific = (value, { type }) => {
    switch (type) {
      case 'date': return <UserTimezoneTimestamp dateTime={value} />;
      case 'boolean': return String(value);
      default: return value;
    }
  };

  render() {
    const { children, field, menuContainer, value, queryId, type } = this.props;
    const element = children || this._renderTypeSpecific(value, type);
    const valueActions = PluginStore.exports('valueActions').map((valueAction) => {
      const onSelect = (event) => {
        this._onMenuToggle();
        valueAction.handler(queryId, event.field, event.value);
      };
      const condition = valueAction.condition || (() => true);
      const actionDisabled = !condition({ field, type, value });
      return (<MenuItem key={`value-action-${field}-${valueAction.type}`}
                        disabled={actionDisabled}
                        eventKey={{ field, value }}
                        onSelect={onSelect}>{valueAction.title}</MenuItem>);
    });

    return (
      <OverlayDropdown show={this.state.open}
                       toggle={element}
                       onToggle={this._onMenuToggle}
                       menuContainer={menuContainer}>
        <div className={style.bottomSpacer}>
          <span className={style.dropdownheader}>
            {field} = {this._renderTypeSpecific(value, type)}
          </span>
        </div>

        <MenuItem divider />
        <MenuItem header>Actions</MenuItem>
        {valueActions}

        <Well className={style.topSpacer}>Found 3827 times in this result set.</Well>
      </OverlayDropdown>
    );
  }
}
