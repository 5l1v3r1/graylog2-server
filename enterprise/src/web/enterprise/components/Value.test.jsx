import React from 'react';
import { mount } from 'enzyme';

import mockComponent from 'helpers/mocking/MockComponent';
import Value from './Value';
import FieldType from '../logic/fieldtypes/FieldType';

jest.mock('./OverlayDropdown', () => mockComponent('OverlayDropdown'));
jest.mock('enterprise/components/common/UserTimezoneTimestamp', () => mockComponent('UserTimezoneTimestamp'));

describe('Value', () => {
  it('render without type information but no children', () => {
    const wrapper = mount(<Value field="foo" queryId="someQueryId" value={42} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders timestamps with a custom component', () => {
    const wrapper = mount(<Value field="foo" queryId="someQueryId" value="2018-10-02T14:45:40Z" type={new FieldType('date', [], [])} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders booleans as strings', () => {
    const wrapper = mount(<Value field="foo" queryId="someQueryId" value={false} type={new FieldType('boolean', [], [])} />);
    expect(wrapper).toMatchSnapshot();
  });
});