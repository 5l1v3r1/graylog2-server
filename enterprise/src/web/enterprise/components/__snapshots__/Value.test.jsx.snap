// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value render without type information but no children 1`] = `
<Value
  field="foo"
  interactive={false}
  menuContainer={<body />}
  queryId="someQueryId"
  type={
    FieldType {
      "value": Immutable.Map {
        "type": "unknown",
        "properties": Immutable.Set [],
        "indexNames": Immutable.Set [],
      },
    }
  }
  value={42}
  viewId={null}
>
  <OverlayDropdown
    menuContainer={<body />}
    onToggle={[Function]}
    show={false}
    toggle={42}
  >
    <overlay-dropdown
      menuContainer={<body />}
      onToggle={[Function]}
      show={false}
      toggle={42}
    >
      <div
        className="bottomSpacer"
      >
        <span
          className="dropdownheader"
        >
          foo
           = 
          42
        </span>
      </div>
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={true}
        header={false}
      >
        <li
          className="divider"
          role="separator"
        />
      </MenuItem>
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={false}
        header={true}
      >
        <li
          className="dropdown-header"
          role="heading"
        >
          Actions
        </li>
      </MenuItem>
    </overlay-dropdown>
  </OverlayDropdown>
</Value>
`;

exports[`Value renders booleans as strings 1`] = `
<Value
  field="foo"
  interactive={false}
  menuContainer={<body />}
  queryId="someQueryId"
  type={
    FieldType {
      "value": Immutable.Map {
        "type": "boolean",
        "properties": Immutable.Set [],
        "indexNames": Immutable.Set [],
      },
    }
  }
  value={false}
  viewId={null}
>
  <OverlayDropdown
    menuContainer={<body />}
    onToggle={[Function]}
    show={false}
    toggle="false"
  >
    <overlay-dropdown
      menuContainer={<body />}
      onToggle={[Function]}
      show={false}
      toggle="false"
    >
      <div
        className="bottomSpacer"
      >
        <span
          className="dropdownheader"
        >
          foo
           = 
          false
        </span>
      </div>
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={true}
        header={false}
      >
        <li
          className="divider"
          role="separator"
        />
      </MenuItem>
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={false}
        header={true}
      >
        <li
          className="dropdown-header"
          role="heading"
        >
          Actions
        </li>
      </MenuItem>
    </overlay-dropdown>
  </OverlayDropdown>
</Value>
`;

exports[`Value renders timestamps with a custom component 1`] = `
<Value
  field="foo"
  interactive={false}
  menuContainer={<body />}
  queryId="someQueryId"
  type={
    FieldType {
      "value": Immutable.Map {
        "type": "date",
        "properties": Immutable.Set [],
        "indexNames": Immutable.Set [],
      },
    }
  }
  value="2018-10-02T14:45:40Z"
  viewId={null}
>
  <OverlayDropdown
    menuContainer={<body />}
    onToggle={[Function]}
    show={false}
    toggle={
      <UserTimezoneTimestamp
        dateTime="2018-10-02T14:45:40Z"
      />
    }
  >
    <overlay-dropdown
      menuContainer={<body />}
      onToggle={[Function]}
      show={false}
      toggle={
        <UserTimezoneTimestamp
          dateTime="2018-10-02T14:45:40Z"
        />
      }
    >
      <div
        className="bottomSpacer"
      >
        <span
          className="dropdownheader"
        >
          foo
           = 
          <UserTimezoneTimestamp
            dateTime="2018-10-02T14:45:40Z"
          >
            <user-timezone-timestamp
              dateTime="2018-10-02T14:45:40Z"
            />
          </UserTimezoneTimestamp>
        </span>
      </div>
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={true}
        header={false}
      >
        <li
          className="divider"
          role="separator"
        />
      </MenuItem>
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={false}
        header={true}
      >
        <li
          className="dropdown-header"
          role="heading"
        >
          Actions
        </li>
      </MenuItem>
    </overlay-dropdown>
  </OverlayDropdown>
</Value>
`;
